### ISSUE TITLE
Inclusion Request: swig

### ISSUE THREAD

**Issue by ss1978** [bincrafters/community#610]
### Package Details
  * Package Name: **swig**
  * Operating Systems+versions: **Linux Ubuntu 16.04, Ubuntu 18.04, Windows, macOS**
  * Compiler support: **gcc-5.4, gcc6, gcc7, clang7**
  * Conan recipe repository: **https://github.com/ss1978/conan-swig**

### Inclusion request
  * Fill all checkboxes with x if you agree with the statement or if it applies to your Conan recipe
  * [x] I want that my Conan recipe is getting included into the Bincrafters organisation
  * [x] I understand that if my request is accepted we are cloning your repository with all its history into Bincrafters, adapt some configurations and meta data if needed, and publishing it under the Bincrafters name and accounts
  * [x] I have used the Bincrafters templates to develop my Conan recipe - https://github.com/bincrafters/conan-templates
  * [x] My Conan recipe and all parts of it are licensed under the MIT license - https://choosealicense.com/licenses/mit/
    * Bincrafters are releasing all recipes under MIT license. If you want to have your recipe within Bincrafters we kindly ask you to do the same
    * it is okay if your test_package contains code from the upstream project under the upstream project's license, as long as it is clearly mentioned within the test_package files
  * [x] My Conan recipe covers (at least) the latest stable version which is available for the package
  * [x] I know that my Conan recipe is working on all operating system and compilers mentioned above right now
    * if not: please provide here links to your CI platforms and / or build logs and describe the problems you encounter
  * [x] I'm willing to help to maintain the recipe even after it is included into Bincrafters

### Checklist for the Bincrafters team
  * Ignore this if you are not part of the Bincrafters team
  * [x] There is no such Conan package yet in conan-center, conan-community or Bincrafters
  * [x] The project is either popular or deserves to be maintained for another good reason (we can't package every single project out there)
  * [x] [bincrafters-conventions](https://github.com/bincrafters/bincrafters-conventions) reports no complains; the recipe fulfills our general quality requirements and conventions
  * [x] The recipe is cloned into our organisation; CI is enabled and configurated and succeeds
  * [x] There is a branch protection rule for `stable/*` branches, which requires a pull request with at least one approval first before things can get published
  * Considering granding the inclusion request author push rights to the recipe repository (there is no guarantee that everyone is getting push rights, we decide this on a case-to-case base)

**Comment by ss1978** [comment]
> Are you in Slack?

Created one, with my current primary email.

**Comment by solvingj** [comment]
I was unable to find you in slack. Please ping me @solvingj

**Comment by madebr** [comment]
Shouldn't this package be renamed to `swig_installer`?

**Comment by uilianries** [comment]
> Shouldn't this package be renamed to `swig_installer`?

if only binaries are provided by this package, so yes.

**Comment by madebr** [comment]
FYI: I've forked the conan-swig repo at https://github.com/madebr/conan-swig

**Comment by madebr** [comment]
I've renamed the repo https://github.com/madebr/conan-swig_installer.
Depends on
- bincrafters/conan-bison#4
- bincrafters/conan-pcre#9

**Comment by Croydon** [comment]
@madebr Could you please remove this custom caching logic in the recipe?

Also in the meantime a new major version got published for swig. Would you be willing to add 4.0?

**Comment by madebr** [comment]
Updated to 4.0.0: https://github.com/madebr/conan-swig_installer/tree/testing/4.0.0

[travis-ci](https://travis-ci.com/madebr/conan-swig_installer/builds/111505050)
[appveyor](https://ci.appveyor.com/project/madebr/conan-swig-installer/builds/24489537)

- The windows version is a repackaged swigwin. It is only available as a 32-bit pre-built executable. The `conanfile.py` removes `self.info.settings.arch_build` on windows so the 32-bit `swig.exe` is used on both `x86` and `x86_64` Windows 
- Builds on 32- and 64-bit `conanio/gcc7-centos6`
- Builds on 64-bit macosx
- Does *not* build on 32-bit macosx due to missing m4_installer and bison_installer

**Comment by madebr** [comment]
I am trying to build swig on Windows instead of using the prebuilt one in the `swigwin` archive.
This builds and runs fine on my Windows machine, but fails on appveyor.
When looking at the output, it downloads [pcre/8.41@bincrafters/stable:338b5858bc652693e9870677b370ca2c38b8d56d](https://ci.appveyor.com/project/madebr/conan-swig-installer/builds/24520295#L325) which is built for [Visual Studio 14 2015](https://ci.appveyor.com/project/BinCrafters/conan-pcre/build/job/t69c9wb5xlsyddpe?fullLog=true#L1906)

If anybody knows how to fix this, please tell me 

conanfile recipe: https://github.com/madebr/conan-swig_installer/tree/win_build
appveyor: https://ci.appveyor.com/project/madebr/conan-swig-installer/builds/24520295

**Comment by madebr** [comment]
@Croydon 

Found the problem.
In `build.py`, I added a new build with only `arch_build` and `os`. This would let conan use the default compiler, which is `Visual Studio` on appveyor.
As a workaround, I copy the `compiler` and `build_requires` from the common builds.
I know this is messy, but I don't know of a better way.

On windows, a 32- and 64-bit build should now be built.
The 32-bit build is not building, because pcre for gcc 32-bit x86 on Windows is missing.

**Comment by Croydon** [comment]
Would it be possible to build Swig with Visual Studio?

**Comment by madebr** [comment]
I haven't tested, but swig does not explain it its [documentation](http://swig.org/Doc4.0/SWIGDocumentation.html#Windows_swig_exe) how to build swig with Visual Studio.
It explains how to use swig in Visual Studio though.

**Comment by madebr** [comment]
btw, I don't think a 32-bit macosx swig will be possible because neither pcre (or any of its dependencies) have a 32-bit version available.

**Comment by Croydon** [comment]
We never build 32-bit macOS builds, this is a non-starter 😃

**Comment by madebr** [comment]
By using [cccl](https://github.com/swig/cccl), I an able to compile swig using Visual Studio.
`ccache-swig` cannot be compiled trivially because it uses posix only includes and libraries.

But since swig is really only an executable and some text files, is it really required to be built using Visual Studio?

**Comment by Croydon** [comment]
No, MinGW is fine in this case

**Comment by madebr** [comment]
I added support for Visual Studio but build the binaries using mingw on appveyor.
I packages cccl at https://github.com/madebr/conan-cccl

swig recipe: https://github.com/madebr/conan-swig_installer/tree/testing/4.0.0
travis-ci: https://travis-ci.com/madebr/conan-swig_installer/builds/112154033
appveyor: https://ci.appveyor.com/project/madebr/conan-swig-installer/builds/24606049

**Comment by Croydon** [comment]
@madebr Thanks, please create an inclusion request for cccl_installer

**Comment by cher-nov** [comment]
@Croydon Any progress on this? Would very appreciate it, as SWIG is used in my university thesis. :C
But thanks anyway.

**Comment by Croydon** [comment]
swig_installer is now available in our Bincrafters repository. Good luck with your thesis @cher-nov :) 

The recipe didn't add the license files to the package, fixed via 
https://github.com/bincrafters/conan-swig_installer/commit/2213f764738e4660f0e2506eaa0ea982efbf4edd 

conan-center inclusion is requested. Still blocked by #823

**Comment by cher-nov** [comment]
@Croydon Tried to install, but got this error. :c
`conans.errors.ConanException: Version range '~=4.0.0' from requirement 'swig_installer/[~=4.0.0]@bincrafters/stable' required by 'virtual' could not be resolved`

And I already got graduated, but thank you anyway. ;)

**Comment by madebr** [comment]
@Croydon I cannot access https://ci.appveyor.com/project/bincrafters/conan-swig_installer

**Comment by Croydon** [comment]
For some reason the AppVeyor URI is https://ci.appveyor.com/project/bincrafters/conan-swig-installer

**Comment by Croydon** [comment]
`conan install swig_installer/[~=4.0.0]@bincrafters/stable`

does work for me

@cher-nov Did you add our Bincrafters repository?

**Comment by cher-nov** [comment]
@Croydon

> Did you add our Bincrafters repository?

Yes.

**Comment by cher-nov** [comment]
@Croydon I'm sorry, my bad. Got misconfigured something since I'm using Conan through its API. Everything seems to work now.

**Comment by ss1978** [comment]
I close this request, since the swig installer is online @ bincrafters repo.
Thanks for the commits and comments.

**Comment by Croydon** [comment]
Reopening until it is in conan-center

**Comment by Croydon** [comment]
Closing now. As with the conan-center-index the process of adding things to conan-center has changed fundamentally.
